django/db/backends/utils.py

```diff
@@ -66,16 +67,26 @@ def callproc(self, procname, params=None, kparams=None):
 
     def execute(self, sql, params=None):
         self.db.validate_no_broken_transaction()
-        with self.db.wrap_database_errors:
-            if params is None:
-                return self.cursor.execute(sql)
-            else:
-                return self.cursor.execute(sql, params)
+        with ExitStack() as stack:
```

Comment:
I'm slightly concerned of the performance implications of *always* invoking an `ExitStack()` for every single query, regardless of whether there are any hooks. It costs about 1.3 microseconds:

```python
In [4]: %timeit with ExitStack(): pass
The slowest run took 10.70 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.33 µs per loop
```

which is fairly significant if you're doing 30 or 100 queries in a view, and want it to return in 100ms.

About half of this cost is creating a new `ExitStack()` instance, I wonder if they can be reused? Also you could make a context manager on the db backend that is a no-op if there are no hooks, and otherwise wraps up the logic duplicated in the two methods here with some kind of caching like `wrap_database_errors` gives us

Reply:
I considered doing this extra contextmanager in order to avoid the repetition, but thought that it would make the code less clear and less readable.

I'm a little confused at why you think the numbers imply a performance issue -- if you execute a 100 queries in your view, and we've added 1.3 microseconds per query, then we've added, altogether, 0.13 milliseconds. I don't consider that significant.